pkg/autoupdate/tuf/library_manager.go

```diff
@@ -0,0 +1,349 @@
+	return filepath.Join(ulm.baseDir, fmt.Sprintf("%s-staged", binary))
+}
+
+// addToLibrary adds the given target file to the library for the given binary,
+// downloading and verifying it if it's not already there.
+func (ulm *updateLibraryManager) AddToLibrary(binary autoupdatableBinary, targetFilename string, targetMetadata data.TargetFileMeta) error {
+	// Check to see if the current running version is the version we were requested to add;
+	// return early if it is, but don't error out if we can't determine the current version.
+	currentVersion, err := ulm.currentRunningVersion(binary)
+	if err != nil {
+		level.Debug(ulm.logger).Log("msg", "could not get current running version", "binary", binary, "err", err)
+	} else if currentVersion == ulm.versionFromTarget(binary, targetFilename) {
+		// We don't need to download the current running version because it already exists,
+		// either in this updates library or in the original install location.
+		return nil
+	}
+
+	if ulm.alreadyAdded(binary, targetFilename) {
+		return nil
+	}
+
+	// Remove downloaded archives after update, regardless of success
+	defer ulm.tidyStagedUpdates(binary)
+
+	stagedUpdatePath, err := ulm.stageAndVerifyUpdate(binary, targetFilename, targetMetadata)
+	if err != nil {
+		return fmt.Errorf("could not stage update: %w", err)
+	}
+
+	if err := ulm.moveVerifiedUpdate(binary, targetFilename, stagedUpdatePath); err != nil {
+		return fmt.Errorf("could not move verified update: %w", err)
+	}
+
+	return nil
+}
+
+// alreadyAdded checks if the given target already exists in the update library.
+func (ulm *updateLibraryManager) alreadyAdded(binary autoupdatableBinary, targetFilename string) bool {
+	updateDirectory := filepath.Join(ulm.updatesDirectory(binary), ulm.versionFromTarget(binary, targetFilename))
+
+	return autoupdate.CheckExecutable(context.TODO(), executableLocation(updateDirectory, binary), "--version") == nil
+}
+
+// versionFromTarget extracts the semantic version for an update from its filename.
+func (ulm *updateLibraryManager) versionFromTarget(binary autoupdatableBinary, targetFilename string) string {
+	// The target is in the form `launcher-0.13.6.tar.gz` -- trim the prefix and the file extension to return the version
+	prefixToTrim := fmt.Sprintf("%s-", binary)
+
+	return strings.TrimSuffix(strings.TrimPrefix(targetFilename, prefixToTrim), ".tar.gz")
+}
+
+// stageUpdate downloads the update indicated by `targetFilename` and verifies it against
+// the given, validated local metadata.
+func (ulm *updateLibraryManager) stageAndVerifyUpdate(binary autoupdatableBinary, targetFilename string, localTargetMetadata data.TargetFileMeta) (string, error) {
+	// Create the staging file
+	stagedUpdatePath := filepath.Join(ulm.stagedUpdatesDirectory(binary), targetFilename)
+	out, err := os.Create(stagedUpdatePath)
+	if err != nil {
+		return "", fmt.Errorf("could not create file at %s: %w", stagedUpdatePath, err)
+	}
+	defer out.Close()
```

Comment:
I think this is fine, but it might make sense to explicitly close before the return and check the error.

Reply:
Updated to a defer func() that logs the error if present -- I think I prefer that over having to call + check out.Close() in the 5 places we could return, and if something's very wrong we should catch it when verifying the executable anyway